hack/generator/pkg/astmodel/type_merger.go

```diff
@@ -0,0 +1,178 @@
+	it := reflect.ValueOf(merger)
+	if it.Kind() != reflect.Func {
+		panic("merger must be a function")
+	}
+
+	mergerType := it.Type()
+
+	if mergerType.NumIn() != 2 && mergerType.NumIn() != 3 {
+		panic("merger must take 2 arguments (Type, Type) or 3 arguments (X, Type, Type)")
+	}
+
+	inOffset := mergerType.NumIn() - 2
+	leftArg := mergerType.In(inOffset + 0)
+	rightArg := mergerType.In(inOffset + 1)
+
+	if !leftArg.AssignableTo(typeInterface) || !rightArg.AssignableTo(typeInterface) {
+		panic("merger must take in types assignable to Type")
+	}
+
+	if mergerType.NumOut() != 2 ||
+		mergerType.Out(0) != typeInterface ||
+		mergerType.Out(1) != errorInterface {
+		panic("merger must return (Type, error)")
+	}
+
+	return it, leftArg, rightArg, inOffset != 0
+}
+
+func (m *TypeMerger) Add(mergeFunc interface{}) {
+	merger, leftArg, rightArg, takesCtx := validateMerger(mergeFunc)
+
+	m.mergers = append(m.mergers,
+		mergerRegistration{
+			left:  leftArg,
+			right: rightArg,
+			merge: reflect.MakeFunc(mergerFuncType, func(args []reflect.Value) []reflect.Value {
+				// we dereference the Type here to the underlying value so that
+				// the merger can take either Type or a specific type.
+				// if it takes Type then the compiler/runtime will convert the value back to a Type
+				ctxValue := args[0].Elem()
+				leftValue := args[1].Elem()
+				rightValue := args[2].Elem()
+				if takesCtx {
+					return merger.Call([]reflect.Value{ctxValue, leftValue, rightValue})
+				} else {
+					return merger.Call([]reflect.Value{leftValue, rightValue})
+				}
+			}).Interface().(MergerFunc),
+		})
+}
+
+func (m *TypeMerger) AddUnordered(mergeFunc interface{}) {
+	merger, leftArg, rightArg, takesCtx := validateMerger(mergeFunc)
+
+	m.mergers = append(m.mergers,
+		mergerRegistration{
+			left:  leftArg,
+			right: rightArg,
+			merge: reflect.MakeFunc(mergerFuncType, func(args []reflect.Value) []reflect.Value {
+				ctxValue := args[0].Elem()
+				leftValue := args[1].Elem()
+				rightValue := args[2].Elem()
+				if takesCtx {
+					return merger.Call([]reflect.Value{ctxValue, leftValue, rightValue})
+				} else {
+					return merger.Call([]reflect.Value{leftValue, rightValue})
+				}
+			}).Interface().(MergerFunc),
+		})
+
+	m.mergers = append(m.mergers,
+		mergerRegistration{
+			left:  rightArg, // flipped
+			right: leftArg,  // flipped
+			merge: reflect.MakeFunc(mergerFuncType, func(args []reflect.Value) []reflect.Value {
+				ctxValue := args[0].Elem()
+				leftValue := args[2].Elem()  // flipped
+				rightValue := args[1].Elem() // flipped
+				if takesCtx {
+					return merger.Call([]reflect.Value{ctxValue, leftValue, rightValue})
+				} else {
+					return merger.Call([]reflect.Value{leftValue, rightValue})
+				}
+			}).Interface().(MergerFunc),
+		})
+}
+
+// Merge merges the two types according to the provided mergers and fallback
+func (m *TypeMerger) Merge(left, right Type, ctx ...interface{}) (Type, error) {
```

Comment:
minor: Depending on usage you could have 2 "overloads" `Merge` and `MergeCtx` one of which takes a `ctx` and the other of which doesn't?

Might be cleaner than this pattern - unsure.

Reply:
I considered this and I do think it would be cleaner overall